### PR TITLE
[WIP] Initial implementation of ModuleInfo for testing

### PR DESCRIPTION
**Pull request by jbschlosser**
This PR contains the initial version of `ModuleInfo` for use in testing modules. It's currently a heavy WIP with a lot missing and incomplete, but includes the following:

* (new file) `torch/testing/_internal/common_modules.py`
  * `ModuleInfo` definition - metadata for each module; probably not complete yet
  * `module_db` - the actual `ModuleInfo` database; currently contains entries for several interesting modules
  * `ModuleInput` - analogous to `SampleInput` from OpInfo; contains `FunctionInput`s for both constructor and forward pass inputs
      * Constructor and forward pass inputs are tied together within a `ModuleInput` because they are likely correlated
  * `FunctionInput` - just contains args and kwargs to pass to a function; might go away
  * `@modules` decorator - analogous to `@ops`; specifies a set of modules to run a test over
      * Still needs proper dtype machinery from `@ops`; ideally don't want to duplicate this
  * Some constants used to keep track of all modules under torch.nn:
      * `MODULE_NAMESPACES` - list of all namespaces containing modules
      * `MODULE_CLASSES` - list of all module class objects
      * `MODULE_CLASS_NAMES` - dict from module class object to nice name (e.g. torch.nn.Linear -> "nn.Linear")
  * Temporarily generates some `ModuleInfo` entries from the old `test/test_module_init.py` stuff to get full coverage for factory kwarg checking (see `test_factory_kwargs` below)
* (new file) `test/test_modules.py`
    * Uses all of the above to define tests over modules
    * Current tests:
        * `test_basics` - tests instantiation, printing, forward / backward with gradient checking (test name will probably change)
        * `test_inplace_variant` - tests the in-place variant vs. out-of-place for modules that have this (e.g. torch.nn.ReLU)
        * `test_pickle_unpickle` - tests pickling and unpickling the module for modules that are pickleable
        * `test_factory_kwargs` - tests passing `device` and `dtype` kwargs to modules and ensures that they affect module parameters / buffers appropriately

Current work:
* Defining sets of inputs to pass to modules is fairly awkward / verbose - is there a nicer way to do this?
* What's the best way to make the `dtype` machinery generic for both `OpInfo` and `ModuleInfo`?
* Still need tests with non-contiguous inputs
* Still need gradient checking
* Stil need some criterion-specific stuff
* Reduce forward / backward pass code duplication across tests
* `reference_fn` stuff isn't fully figured out yet - ideally want to be able to specify reference function for each particular case

### CODE REVIEW

**Comment by jbschlosser** [comment]
Closing as unneeded.
